File: lib/Scraper.js
import Knex from 'knex';
import fs from 'fs';
import requestRaw from 'request';
import Promise from 'bluebird';
import $ from 'cheerio';
import moment from 'moment';
import slug from 'slug';
import util from 'util';
import { Iconv } from 'iconv';

const request = Promise.promisify(requestRaw);
const USER_AGENT = 'pollofpolls-scraper | https://github.com/jarib/pollofpolls-scraper';
const DATE = moment().format('YYYY-MM-DD');

export default class Scraper {
    static run(config) {
        (new this(config)).run();
    }

    constructor(config) {
        const { db } = config;

        if (fs.existsSync(db)) {
            fs.unlinkSync(db);
        }

        this.knex = Knex({
            debug: false,
            client: 'sqlite3',
            connection: { filename: db }
        });

        this.iconv = new Iconv('ISO-8859-1', 'UTF-8');
    }

    run() {
        this.scrapeMunicipalityElection()
            .then(::this.scrapeCountyElection)
            .then(::this.scrapeParliamentElection)
            .finally(::this.done)
    }

    done() {
        this.knex.destroy();
    }

    scrapeMunicipalityElection() {
        return this
            .fetch(`http://www.pollofpolls.no/?cmd=Kommunestyre&do=vispopalle`)
            .then(::this.parseTable)
            .then(this.save.bind(this, 'municipality'))
    }

    scrapeCountyElection() {
        return this
            .fetch('http://www.pollofpolls.no/?cmd=Fylkesting&do=vispopalle')
            .then(::this.parseTable)
            .then(this.save.bind(this, 'county'))

    }

    scrapeParliamentElection() {
        return this
            .fetch(`http://www.pollofpolls.no/?cmd=Stortinget&do=vispopalle`)
            .then(::this.parseTable)
            .then(this.save.bind(this, 'parliament'))

    }

    parseTable(doc) {
        const result = []

        doc('#content table tr').each((i, row) => {
            let data = []

            $('th, td', row).each(
                (i, cell) => data.push($(cell).text())
            );

            result.push(data);
        });

        const [ header, ...rows ] = result;
        return { header, rows };
    }

    save(election, data) {
        const parties = data.header.slice(1).map(e => slug(e, {lower: true, replacement: '_'}));

        return this.createSchema(election, parties)
            .then(() => {
                return Promise.each(data.rows, row => {
                    const [name, ...data] = row
                    const results = {};

                    if (data.length !== parties.length) {
                        console.error(`skipping invalid row: ${util.inspect(row)}`)
                        return;
                    }

                    const date = this.dateFrom(name);

                    if (!date) {
                        console.error(`unable to parse date from name: ${name}`)
                    }

                    parties.forEach((slug, index) => {
                        var m = data[index].match(/^([\d,]+) \((\d+)\)/)

                        if (!m) {
                            console.error(`unable to parse cell: ${data[index]}`)
                        }

                        results[`${slug}_percent`] = +(m[1].replace(',', '.'))
                        results[`${slug}_mandates`] = +(m[2]);
                    })

                    return this.knex(election).insert({
                        name,
                        date,
                        ...results
                    });
                });
            });
    }

    dateFrom(name) {
        const m = name.match(/^Uke (\d+)-(\d+)/);

        if (!m) {
            return null;
        }

        return moment(`${m[1]}-${m[2]}`, 'W-YYYY').endOf('week').format('YYYY-MM-DD');
    }

    fetch(url) {
        return request({
            url: url,
            headers: { 'User-Agent': USER_AGENT },
            encoding: null
        }).spread((res, body) => {
            if (res.statusCode !== 200) {
                throw new Error(`request failed - ${res.statusCode}: ${res.body}`)
            }

            return $.load(this.iconv.convert(body).toString());
        });
    }

    createSchema(tableName, parties) {
        return this.knex.schema.createTable(tableName, table => {
            table.string('name');
            table.date('date');

            parties.forEach(party => {
                table.float(`${party}_percent`);
                table.integer(`${party}_mandates`);
            });
        });
    }

}

File: scraper.rb
require "faraday"
require "nokogiri"
require "sqlite3"
require "fileutils"
require "date"
require "pry"

class Scraper

  PARTIES = {
    'Ap'            => 'A',
    'Høyre'         => 'H',
    'H'             => 'H',
    'Venstre'       => 'V',
    'Krf'           => 'KrF',
    'KrF'           => 'KrF',
    'Frp'           => 'FrP',
    'Rødt'          => 'R',
    'MDG'           => 'MDG',
    'SV'            => 'SV',
    'Sp'            => 'Sp',
    'V'             => 'V',
    'Andre'         => 'Andre',
    'Andre partier' => 'Andre',
  }

  def initialize(path)
    FileUtils.rm_rf path

    @db = SQLite3::Database.new path
    @db.trace { |sql| p sql } if $DEBUG
  end

  def run
    create_table

    scrape_pop_municipality
    scrape_pop_county
    scrape_pop_parliament
    scrape_pop_snitt

    # scrape_infact
  end

  private

  def create_table
    @db.execute <<-SQL
    CREATE TABLE polls (
      startDate date,
      endDate date NOT NULL,
      source varchar(255) NOT NULL,
      election varchar(50),
      region varchar(255) NOT NULL,
      party varchar(10) NOT NULL,
      percentage float NOT NULL,
      comment varchar(255),
      mandates integer
    )
    SQL
  end

  def scrape_pop_municipality
    save(
      parse(fetch('http://www.pollofpolls.no/?cmd=Kommunestyre&do=vispopalle&dato=2015-08-22')),
      source: 'pollofpolls.no',
      election: 'municipality',
      region: 'Norge'
    )
  end

  def scrape_pop_county
    save(
      parse(fetch("http://www.pollofpolls.no/?cmd=Fylkesting&do=vispopalle")),
      source: 'pollofpolls.no',
      election: 'county',
      region: 'Norge'
    )
  end

  def scrape_pop_parliament
    save(
      parse(fetch("http://www.pollofpolls.no/?cmd=Stortinget&do=vispopalle")),
      source: 'pollofpolls.no',
      election: 'parliament',
      region: 'Norge'
    )
  end

  def scrape_pop_snitt
    save(
      parse(fetch("http://www.pollofpolls.no/?cmd=Stortinget&do=visallesnitt")),
      source: 'pollofpolls.no',
      election: 'snitt',
      region: 'Norge'
    )
  end

  MONTHS = {
    'Jan' => 1,
    'Januar' => 1,
    'Feb' => 2,
    'Februar' => 2,
    'Mars' => 3,
    'April' => 4,
    'Mai' => 5,
    'Juni' => 6,
    'Juli' => 7,
    'Aug' => 8,
    'August' => 8,
    'Sept' => 9,
    'September' => 9,
    'Okt' => 10,
    'Oktober' => 10,
    'Nov' => 11,
    'November' => 11,
    'Des' => 12,
    'Desember' => 12
  }

  def fetch(url)
    body = Faraday.get(url).body.force_encoding('UTF-8')
  end

  def parse(str)
    doc = Nokogiri::HTML.parse(str)

    rows = []

    doc.css('#content table tr').each do |row|
      rows << row.css('th, td').map { |e| e.text.strip }
    end

    {
      header: rows.shift,
      rows: rows
    }
  end

  def save(data, opts = {})
    parties = data[:header][1..-1].map { |e| PARTIES.fetch(e) }

    previous_row_name = nil

    data[:rows].each do |row|
      row_name, *cells = row

      if cells.size != parties.size
        puts "skipping invalid row: #{row.inspect}"
        next
      end

      # fix PoP bugs
      if row_name == 'Uke 1-2014' && previous_row_name == "Uke 2-2015"
        row_name = "Uke 1-2015"
      elsif row_name == 'Uke 1-2013' && previous_row_name == "Uke 2-2014"
        row_name = "Uke 1-2014"
      end

      start_date, end_date = dates_from(row_name)


      unless end_date
        puts "unable to parse date from name: #{row_name.inspect}"
        next
      end

      cells.each_with_index do |cell, idx|
        if cell =~ /^([\d,]+) \((\d+)\)/
          percent, mandates = $1, $2

          cols = {
            comment: row_name,
            startDate: start_date.strftime("%Y-%m-%d"),
            endDate: end_date.strftime("%Y-%m-%d"),
            percentage: Float(percent.sub(',', '.')),
            mandates: Integer(mandates),
            party: parties[idx]
          }.merge(opts)

          save_row cols
        else
          raise "unable to parse cell: #{cell}"
        end
      end

      previous_row_name = row_name
    end
  end

  def save_row(row)
    keys         = row.keys
    values       = row.keys.map { |k| row[k] }
    placeholders = Array.new(values.size, "?").join(', ')

    sql = "INSERT INTO polls (#{keys.join(', ')}) VALUES ( #{placeholders} )"
    p sql if $DEBUG

    @db.execute(sql, values)
  end

  def dates_from(name)
    p name: name

    if name =~ /^Uke (\d+)-(\d+)/
      eow = Date.strptime("#{$1}-#{$2}", "%U-%Y")
      sow = eow - 6

      [sow, eow]
    elsif name =~ /(#{MONTHS.keys.join('|')}) '(\d{2})/
      month_name = $1
      short_year = $2


      month = MONTHS[month_name] or raise "unknown month #{month_name}"
      year = Integer("20#{short_year}")

      start = Date.new(year, month)

      [start, start.next_month - 1]
    end
  end
end

if __FILE__ == $0
  Scraper.new("data.sqlite").run
end
